File: bindings/nodejs/lib/platform.js
/*
 * FastQR - Fast QR Code Generator Library
 * Copyright (C) 2025 Tran Huu Canh and FastQR Contributors
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, see <https://www.gnu.org/licenses/>.
 *
 * Homepage: https://github.com/tranhuucanh/fastqr
 */

const os = require('os');
const path = require('path');
const fs = require('fs');
const { execSync } = require('child_process');

/**
 * Detects the current platform and architecture
 * @returns {string} Platform identifier (e.g., 'macos-arm64', 'linux-x86_64')
 */
function detectPlatform() {
  const platform = os.platform();
  const arch = os.arch();

  if (platform === 'darwin') {
    if (arch === 'arm64') return 'macos-arm64';
    if (arch === 'x64') return 'macos-x86_64';
  } else if (platform === 'linux') {
    if (arch === 'x64') return 'linux-x86_64';
    if (arch === 'arm64') return 'linux-aarch64';
  }

  throw new Error(`Unsupported platform: ${platform}-${arch}`);
}

/**
 * Extracts pre-built binary from tarball
 * @param {string} tarballPath - Path to the tarball
 * @param {string} destDir - Destination directory
 */
function extractBinary(tarballPath, destDir) {
  if (!fs.existsSync(destDir)) {
    fs.mkdirSync(destDir, { recursive: true });
  }

  execSync(`tar -xzf "${tarballPath}" -C "${destDir}"`, { stdio: 'inherit' });
}

/**
 * Finds the fastqr binary in prebuilt directory
 * @returns {string} Path to fastqr binary
 */
function findFastQRBinary() {
  const platform = detectPlatform();
  const prebuiltDir = path.join(__dirname, '..', 'prebuilt', platform);
  const binaryPath = path.join(prebuiltDir, 'bin', 'fastqr');

  if (platform.startsWith('linux')) {
    // Linux: Check for AppImage first
    if (fs.existsSync(binaryPath) && fs.accessSync(binaryPath, fs.constants.X_OK)) {
      try {
        // Test if AppImage can run (version check)
        const output = execSync(`${binaryPath} -v`, { encoding: 'utf8', stdio: 'pipe' });
        if (output.includes('FastQR')) {
          return binaryPath;
        }
      } catch (error) {
        // AppImage failed, continue to fallback
      }
    }
  } else {
    // macOS/Windows: Use regular binary
    if (fs.existsSync(binaryPath)) {
      fs.chmodSync(binaryPath, 0o755);
      return binaryPath;
    }
  }

  // Try to extract from tarball
  const tarballPath = path.join(__dirname, '..', 'prebuilt', `${platform}.tar.gz`);
  if (fs.existsSync(tarballPath)) {
    console.log(`Extracting pre-built binary from ${tarballPath}...`);
    extractBinary(tarballPath, prebuiltDir);

    if (fs.existsSync(binaryPath)) {
      fs.chmodSync(binaryPath, 0o755);
      return binaryPath;
    }
  }

  throw new Error(`Pre-built binary not found for ${platform}`);
}

/**
 * Checks if pre-built binary is available
 * @returns {boolean}
 */
function isPrebuiltAvailable() {
  try {
    const platform = detectPlatform();
    const prebuiltDir = path.join(__dirname, '..', 'prebuilt', platform, 'bin');
    const binaryPath = path.join(prebuiltDir, 'fastqr');
    return fs.existsSync(binaryPath);
  } catch (error) {
    return false;
  }
}

/**
 * Gets platform string
 * @returns {string}
 */
function getPlatformString() {
  return detectPlatform();
}

module.exports = {
  detectPlatform,
  extractBinary,
  findFastQRBinary,
  isPrebuiltAvailable,
  getPlatformString
};

File: bindings/ruby/lib/fastqr/platform.rb
# frozen_string_literal: true

module FastQR
  # Platform detection and pre-built binary management
  module Platform
    class << self
      # Detects the current platform
      # @return [String] Platform identifier (e.g., 'macos-arm64', 'linux-x86_64')
      def detect
        os = RbConfig::CONFIG['host_os']
        arch = RbConfig::CONFIG['host_cpu']

        case os
        when /darwin/
          case arch
          when /arm64|aarch64/
            'macos-arm64'
          when /x86_64|x64/
            'macos-x86_64'
          else
            raise "Unsupported macOS architecture: #{arch}"
          end
        when /linux/
          case arch
          when /x86_64|x64/
            'linux-x86_64'
          when /arm64|aarch64/
            'linux-aarch64'
          else
            raise "Unsupported Linux architecture: #{arch}"
          end
        else
          raise "Unsupported platform: #{os}"
        end
      end

      # Check if pre-built binary is available
      # @return [Boolean] true if binary exists
      def prebuilt_available?
        binary_path = find_binary
        binary_path && File.exist?(binary_path)
      rescue StandardError
        false
      end

      # Get path to the library file
      # @return [String] Path to libfastqr.dylib or libfastqr.so
      def lib_path
        platform = detect
        base_dir = File.expand_path('../../prebuilt', __dir__)
        lib_file = platform.start_with?('macos') ? 'libfastqr.dylib' : 'libfastqr.so'
        File.join(base_dir, platform, 'lib', lib_file)
      end

      # Get binary name
      # @return [String] 'fastqr'
      def binary_name
        'fastqr'
      end

      # Extracts pre-built binary from tarball
      # @param tarball_path [String] Path to the tarball
      # @param dest_dir [String] Destination directory
      def extract_binary(tarball_path, dest_dir)
        FileUtils.mkdir_p(dest_dir)
        system("tar -xzf '#{tarball_path}' -C '#{dest_dir}'") or raise "Failed to extract #{tarball_path}"
      end

      # Finds the fastqr binary
      # @return [String] Path to fastqr binary
      def find_binary
        platform = detect
        base_dir = File.expand_path('../../prebuilt', __dir__)
        binary_path = File.join(base_dir, platform, 'bin', 'fastqr')

        if platform.start_with?('linux')
          # Linux: Check for AppImage first
          if File.exist?(binary_path) && File.executable?(binary_path)
            # Test if AppImage can run (version check)
            begin
              output = `#{binary_path} -v 2>&1`.strip
              if $?.success? && output.include?('FastQR')
                return binary_path
              end
            rescue => e
              # AppImage failed, continue to fallback
            end
          end
        else
          # macOS/Windows: Use regular binary
          return binary_path if File.exist?(binary_path)
        end

        # Try to extract from tarball
        tarball_path = File.join(base_dir, "#{platform}.tar.gz")
        if File.exist?(tarball_path)
          puts "Extracting pre-built binary from #{tarball_path}..."
          extract_binary(tarball_path, File.join(base_dir, platform))

          if File.exist?(binary_path)
            File.chmod(0755, binary_path)
            return binary_path
          end
        end

        raise "Pre-built binary not found for #{platform}"
      end
    end
  end
end
